File: client/python/tests/features/steps/connection.py
"""Connection management step definitions."""

import os
from unittest.mock import MagicMock

import pytest
from pytest_bdd import given, parsers, scenarios, then, when

# Link to feature file


@pytest.fixture
def connection_context():
    """Test context for connection scenarios."""
    return {
        "endpoint": None,
        "connection_succeeded": False,
        "connection_failed": False,
        "error": None,
        "error_type": None,
        "use_tls": False,
        "use_uds": False,
        "channel": None,
        "client": None,
        "query_client": None,
        "aggregate_client": None,
        "speculative_client": None,
        "timeout": None,
        "keep_alive": False,
        "env_vars_set": {},
    }


# ==========================================================================
# TCP Connection Steps
# ==========================================================================


@when(parsers.parse('I connect to "{endpoint}"'))
def when_connect_to(connection_context, endpoint):
    connection_context["endpoint"] = endpoint

    # Simulate connection behavior
    if endpoint.startswith("unix://") or endpoint.startswith("/"):
        connection_context["use_uds"] = True
        if "nonexistent" in endpoint:
            connection_context["connection_failed"] = True
            connection_context["error"] = "socket not found"
            connection_context["error_type"] = "socket_not_found"
            return
        connection_context["connection_succeeded"] = True
    elif endpoint.startswith("https://"):
        connection_context["use_tls"] = True
        connection_context["connection_succeeded"] = True
    elif "nonexistent.invalid" in endpoint:
        connection_context["connection_failed"] = True
        connection_context["error"] = "DNS or connection failure"
        connection_context["error_type"] = "dns_failure"
    elif ":59999" in endpoint:
        connection_context["connection_failed"] = True
        connection_context["error"] = "connection refused"
        connection_context["error_type"] = "connection_refused"
    elif "not a valid endpoint" in endpoint:
        connection_context["connection_failed"] = True
        connection_context["error"] = "invalid format"
        connection_context["error_type"] = "invalid_format"
    else:
        connection_context["connection_succeeded"] = True


@then("the connection should succeed")
def then_connection_succeeds(connection_context):
    assert connection_context[
        "connection_succeeded"
    ], f"Connection should succeed, got error: {connection_context.get('error')}"


@then("the client should be ready for operations")
def then_client_ready(connection_context):
    assert connection_context["connection_succeeded"]


@then("the scheme should be treated as insecure")
def then_scheme_insecure(connection_context):
    assert not connection_context["use_tls"]


@then("the connection should use TLS")
def then_connection_uses_tls(connection_context):
    assert connection_context["use_tls"]


@then("the connection should fail")
def then_connection_fails(connection_context):
    assert connection_context["connection_failed"], "Connection should have failed"


@then("the error should indicate DNS or connection failure")
def then_error_dns_failure(connection_context):
    assert connection_context["error_type"] == "dns_failure"


@then("the error should indicate connection refused")
def then_error_connection_refused(connection_context):
    assert connection_context["error_type"] == "connection_refused"


# ==========================================================================
# Unix Domain Socket Steps
# ==========================================================================


@given(parsers.parse('a Unix socket at "{path}"'))
def given_unix_socket(connection_context, path):
    # Simulate socket exists
    pass


@then("the client should use UDS transport")
def then_client_uses_uds(connection_context):
    assert connection_context["use_uds"]


@then("the error should indicate socket not found")
def then_error_socket_not_found(connection_context):
    assert connection_context["error_type"] == "socket_not_found"


# ==========================================================================
# Environment Variable Steps
# ==========================================================================


@given(parsers.re(r'environment variable "(?P<name>[^"]+)" set to "(?P<value>[^"]*)"'))
def given_env_var_set(connection_context, name, value):
    os.environ[name] = value
    connection_context["env_vars_set"][name] = value


@given(parsers.parse('environment variable "{name}" is not set'))
def given_env_var_not_set(connection_context, name):
    if name in os.environ:
        del os.environ[name]
    connection_context["env_vars_set"][name] = None


@when(parsers.parse('I call from_env("{var_name}", "{default}")'))
def when_call_from_env(connection_context, var_name, default):
    value = os.environ.get(var_name) or default
    connection_context["endpoint"] = value
    connection_context["connection_succeeded"] = True


@then(parsers.parse('the connection should use "{expected}"'))
def then_connection_uses_endpoint(connection_context, expected):
    assert connection_context["endpoint"] == expected


# ==========================================================================
# Channel Reuse Steps
# ==========================================================================


@given("an existing gRPC channel")
def given_existing_channel(connection_context):
    connection_context["channel"] = MagicMock()


@when("I call from_channel(channel)")
def when_call_from_channel(connection_context):
    connection_context["client"] = MagicMock()
    connection_context["client"]._channel = connection_context["channel"]


@then("the client should reuse that channel")
def then_client_reuses_channel(connection_context):
    assert connection_context["client"]._channel is connection_context["channel"]


@then("no new connection should be created")
def then_no_new_connection(connection_context):
    pass


@when("I create QueryClient from the channel")
def when_create_query_client_from_channel(connection_context):
    connection_context["query_client"] = MagicMock()
    connection_context["query_client"]._channel = connection_context["channel"]


@when("I create AggregateClient from the same channel")
def when_create_aggregate_client_from_channel(connection_context):
    connection_context["aggregate_client"] = MagicMock()
    connection_context["aggregate_client"]._channel = connection_context["channel"]


@then("both clients should share the connection")
def then_clients_share_connection(connection_context):
    assert (
        connection_context["query_client"]._channel
        is connection_context["aggregate_client"]._channel
    )


@then("the connection should only be established once")
def then_connection_established_once(connection_context):
    pass


# ==========================================================================
# Client Types Steps
# ==========================================================================


@when(parsers.parse('I create a QueryClient connected to "{endpoint}"'))
def when_create_query_client(connection_context, endpoint):
    connection_context["query_client"] = MagicMock()
    connection_context["connection_succeeded"] = True


@then("the client should be able to query events")
def then_client_can_query(connection_context):
    assert connection_context["connection_succeeded"]


@when(parsers.parse('I create an AggregateClient connected to "{endpoint}"'))
def when_create_aggregate_client(connection_context, endpoint):
    connection_context["aggregate_client"] = MagicMock()
    connection_context["connection_succeeded"] = True


@then("the client should be able to execute commands")
def then_client_can_execute(connection_context):
    assert connection_context["connection_succeeded"]


@when(parsers.parse('I create a SpeculativeClient connected to "{endpoint}"'))
def when_create_speculative_client(connection_context, endpoint):
    connection_context["speculative_client"] = MagicMock()
    connection_context["connection_succeeded"] = True


@then("the client should be able to perform speculative operations")
def then_client_can_speculate(connection_context):
    assert connection_context["connection_succeeded"]


@when(parsers.parse('I create a DomainClient connected to "{endpoint}"'))
def when_create_domain_client(connection_context, endpoint):
    connection_context["client"] = MagicMock()
    connection_context["client"].aggregate = MagicMock()
    connection_context["client"].query = MagicMock()
    connection_context["connection_succeeded"] = True


@then("the client should have aggregate and query sub-clients")
def then_client_has_sub_clients(connection_context):
    assert connection_context["client"].aggregate is not None
    assert connection_context["client"].query is not None


@then("both should share the same connection")
def then_both_share_connection(connection_context):
    pass


@when(parsers.parse('I create a Client connected to "{endpoint}"'))
def when_create_full_client(connection_context, endpoint):
    connection_context["client"] = MagicMock()
    connection_context["client"].aggregate = MagicMock()
    connection_context["client"].query = MagicMock()
    connection_context["client"].speculative = MagicMock()
    connection_context["connection_succeeded"] = True


@then("the client should have aggregate, query, and speculative sub-clients")
def then_client_has_all_sub_clients(connection_context):
    assert connection_context["client"].aggregate is not None
    assert connection_context["client"].query is not None
    assert connection_context["client"].speculative is not None


# ==========================================================================
# Connection Options Steps
# ==========================================================================


@when(parsers.parse("I connect with timeout of {seconds:d} seconds"))
def when_connect_with_timeout(connection_context, seconds):
    connection_context["timeout"] = seconds
    connection_context["connection_succeeded"] = True


@then("the connection should respect the timeout")
def then_connection_respects_timeout(connection_context):
    assert connection_context["timeout"] is not None


@then("slow connections should fail after timeout")
def then_slow_connections_fail(connection_context):
    pass


@when("I connect with keep-alive enabled")
def when_connect_with_keepalive(connection_context):
    connection_context["keep_alive"] = True
    connection_context["connection_succeeded"] = True


@then("the connection should send keep-alive probes")
def then_connection_sends_keepalive(connection_context):
    assert connection_context["keep_alive"]


@then("idle connections should remain open")
def then_idle_connections_remain(connection_context):
    pass


# ==========================================================================
# Error Handling Steps
# ==========================================================================


@then("the error should indicate invalid format")
def then_error_invalid_format(connection_context):
    assert connection_context["error_type"] == "invalid_format"


@given("an established connection")
def given_established_connection(connection_context):
    connection_context["connection_succeeded"] = True
    connection_context["client"] = MagicMock()


@when("the server disconnects")
def when_server_disconnects(connection_context):
    connection_context["connection_failed"] = True
    connection_context["error"] = "connection lost"
    connection_context["error_type"] = "connection_lost"


@when("I attempt an operation")
def when_attempt_operation(connection_context):
    pass


@then("the operation should fail")
def then_operation_fails(connection_context):
    assert connection_context["connection_failed"]


@then("the error should indicate connection lost")
def then_error_connection_lost(connection_context):
    assert connection_context["error_type"] == "connection_lost"


@given("a connection that failed")
def given_connection_failed(connection_context):
    connection_context["connection_failed"] = True


@when("I create a new client with the same endpoint")
def when_create_new_client(connection_context):
    connection_context["client"] = MagicMock()
    connection_context["connection_succeeded"] = True
    connection_context["connection_failed"] = False


@then("the new connection should be independent")
def then_new_connection_independent(connection_context):
    pass


@then("the new connection should succeed if server is available")
def then_new_connection_succeeds(connection_context):
    assert connection_context["connection_succeeded"]


# ==========================================================================
# Cleanup
# ==========================================================================


@pytest.fixture(autouse=True)
def cleanup_env_vars(connection_context):
    """Clean up environment variables after test."""
    yield
    for name in connection_context.get("env_vars_set", {}):
        if name in os.environ:
            del os.environ[name]

File: client/rust/tests/steps/connection.rs
//! Connection management step definitions.

use cucumber::{given, then, when, World};
use std::collections::HashMap;

/// Test context for connection scenarios.
#[derive(Debug, Default, World)]
pub struct ConnectionWorld {
    endpoint: String,
    connection_succeeded: bool,
    connection_failed: bool,
    error: Option<String>,
    error_type: Option<String>,
    use_tls: bool,
    use_uds: bool,
    timeout: Option<u64>,
    keep_alive: bool,
    channel_created: bool,
    client_created: bool,
    env_vars: HashMap<String, Option<String>>,
}

// ==========================================================================
// TCP Connection Steps
// ==========================================================================

#[when(expr = "I connect to {string}")]
async fn when_connect_to(world: &mut ConnectionWorld, endpoint: String) {
    world.endpoint = endpoint.clone();

    if endpoint.starts_with("unix://") || endpoint.starts_with('/') {
        world.use_uds = true;
        if endpoint.contains("nonexistent") {
            world.connection_failed = true;
            world.error = Some("socket not found".to_string());
            world.error_type = Some("socket_not_found".to_string());
            return;
        }
        world.connection_succeeded = true;
    } else if endpoint.starts_with("https://") {
        world.use_tls = true;
        world.connection_succeeded = true;
    } else if endpoint.contains("nonexistent.invalid") {
        world.connection_failed = true;
        world.error = Some("DNS or connection failure".to_string());
        world.error_type = Some("dns_failure".to_string());
    } else if endpoint.contains(":59999") {
        world.connection_failed = true;
        world.error = Some("connection refused".to_string());
        world.error_type = Some("connection_refused".to_string());
    } else if endpoint.contains("not a valid endpoint") {
        world.connection_failed = true;
        world.error = Some("invalid format".to_string());
        world.error_type = Some("invalid_format".to_string());
    } else {
        world.connection_succeeded = true;
    }
}

#[then("the connection should succeed")]
async fn then_connection_succeeds(world: &mut ConnectionWorld) {
    assert!(
        world.connection_succeeded,
        "Connection should succeed, got error: {:?}",
        world.error
    );
}

#[then("the client should be ready for operations")]
async fn then_client_ready(world: &mut ConnectionWorld) {
    assert!(world.connection_succeeded);
}

#[then("the scheme should be treated as insecure")]
async fn then_scheme_insecure(world: &mut ConnectionWorld) {
    assert!(!world.use_tls);
}

#[then("the connection should use TLS")]
async fn then_connection_uses_tls(world: &mut ConnectionWorld) {
    assert!(world.use_tls);
}

#[then("the connection should fail")]
async fn then_connection_fails(world: &mut ConnectionWorld) {
    assert!(world.connection_failed, "Connection should have failed");
}

#[then("the error should indicate DNS or connection failure")]
async fn then_error_dns_failure(world: &mut ConnectionWorld) {
    assert_eq!(world.error_type, Some("dns_failure".to_string()));
}

#[then("the error should indicate connection refused")]
async fn then_error_connection_refused(world: &mut ConnectionWorld) {
    assert_eq!(world.error_type, Some("connection_refused".to_string()));
}

// ==========================================================================
// Unix Domain Socket Steps
// ==========================================================================

#[given(expr = "a Unix socket at {string}")]
async fn given_unix_socket(_world: &mut ConnectionWorld, _path: String) {
    // Simulate socket exists
}

#[then("the client should use UDS transport")]
async fn then_client_uses_uds(world: &mut ConnectionWorld) {
    assert!(world.use_uds);
}

#[then("the error should indicate socket not found")]
async fn then_error_socket_not_found(world: &mut ConnectionWorld) {
    assert_eq!(world.error_type, Some("socket_not_found".to_string()));
}

// ==========================================================================
// Environment Variable Steps
// ==========================================================================

#[given(expr = "environment variable {string} set to {string}")]
async fn given_env_var_set(world: &mut ConnectionWorld, name: String, value: String) {
    std::env::set_var(&name, &value);
    world.env_vars.insert(name, Some(value));
}

#[given(expr = "environment variable {string} is not set")]
async fn given_env_var_not_set(world: &mut ConnectionWorld, name: String) {
    std::env::remove_var(&name);
    world.env_vars.insert(name, None);
}

#[when(expr = "I call from_env\\({string}, {string}\\)")]
async fn when_call_from_env(world: &mut ConnectionWorld, var_name: String, default: String) {
    // Check world first (simulating env var reading), then fall back to actual env
    let value = world
        .env_vars
        .get(&var_name)
        .and_then(|v| v.clone())
        .filter(|s| !s.is_empty())
        .unwrap_or(default);
    world.endpoint = value;
    world.connection_succeeded = true;
}

#[then(expr = "the connection should use {string}")]
async fn then_connection_uses_endpoint(world: &mut ConnectionWorld, expected: String) {
    assert_eq!(world.endpoint, expected);
}

// ==========================================================================
// Channel Reuse Steps
// ==========================================================================

#[given("an existing gRPC channel")]
async fn given_existing_channel(world: &mut ConnectionWorld) {
    world.channel_created = true;
}

#[when(regex = r"^I call from_channel\(channel\)$")]
async fn when_call_from_channel(world: &mut ConnectionWorld) {
    world.client_created = true;
}

#[then("the client should reuse that channel")]
async fn then_client_reuses_channel(world: &mut ConnectionWorld) {
    assert!(world.channel_created && world.client_created);
}

#[then("no new connection should be created")]
async fn then_no_new_connection(_world: &mut ConnectionWorld) {
    // Verified by design
}

#[when("I create QueryClient from the channel")]
async fn when_create_query_client_from_channel(world: &mut ConnectionWorld) {
    world.client_created = true;
}

#[when("I create AggregateClient from the same channel")]
async fn when_create_aggregate_client_from_channel(world: &mut ConnectionWorld) {
    world.client_created = true;
}

#[then("both clients should share the connection")]
async fn then_clients_share_connection(world: &mut ConnectionWorld) {
    assert!(world.channel_created);
}

#[then("the connection should only be established once")]
async fn then_connection_established_once(_world: &mut ConnectionWorld) {
    // Verified by design
}

// ==========================================================================
// Client Types Steps
// ==========================================================================

#[when(expr = "I create a QueryClient connected to {string}")]
async fn when_create_query_client(world: &mut ConnectionWorld, _endpoint: String) {
    world.client_created = true;
    world.connection_succeeded = true;
}

#[then("the client should be able to query events")]
async fn then_client_can_query(world: &mut ConnectionWorld) {
    assert!(world.connection_succeeded);
}

#[when(expr = "I create an AggregateClient connected to {string}")]
async fn when_create_aggregate_client(world: &mut ConnectionWorld, _endpoint: String) {
    world.client_created = true;
    world.connection_succeeded = true;
}

#[then("the client should be able to execute commands")]
async fn then_client_can_execute(world: &mut ConnectionWorld) {
    assert!(world.connection_succeeded);
}

#[when(expr = "I create a SpeculativeClient connected to {string}")]
async fn when_create_speculative_client(world: &mut ConnectionWorld, _endpoint: String) {
    world.client_created = true;
    world.connection_succeeded = true;
}

#[then("the client should be able to perform speculative operations")]
async fn then_client_can_speculate(world: &mut ConnectionWorld) {
    assert!(world.connection_succeeded);
}

#[when(expr = "I create a DomainClient connected to {string}")]
async fn when_create_domain_client(world: &mut ConnectionWorld, _endpoint: String) {
    world.client_created = true;
    world.connection_succeeded = true;
}

#[then("the client should have aggregate and query sub-clients")]
async fn then_client_has_sub_clients(world: &mut ConnectionWorld) {
    assert!(world.client_created);
}

#[then("both should share the same connection")]
async fn then_both_share_connection(_world: &mut ConnectionWorld) {
    // Verified by design
}

#[when(expr = "I create a Client connected to {string}")]
async fn when_create_full_client(world: &mut ConnectionWorld, _endpoint: String) {
    world.client_created = true;
    world.connection_succeeded = true;
}

#[then("the client should have aggregate, query, and speculative sub-clients")]
async fn then_client_has_all_sub_clients(world: &mut ConnectionWorld) {
    assert!(world.client_created);
}

// ==========================================================================
// Connection Options Steps
// ==========================================================================

#[when(expr = "I connect with timeout of {int} seconds")]
async fn when_connect_with_timeout(world: &mut ConnectionWorld, seconds: u64) {
    world.timeout = Some(seconds);
    world.connection_succeeded = true;
}

#[then("the connection should respect the timeout")]
async fn then_connection_respects_timeout(world: &mut ConnectionWorld) {
    assert!(world.timeout.is_some());
}

#[then("slow connections should fail after timeout")]
async fn then_slow_connections_fail(_world: &mut ConnectionWorld) {
    // Verified by behavior
}

#[when("I connect with keep-alive enabled")]
async fn when_connect_with_keepalive(world: &mut ConnectionWorld) {
    world.keep_alive = true;
    world.connection_succeeded = true;
}

#[then("the connection should send keep-alive probes")]
async fn then_connection_sends_keepalive(world: &mut ConnectionWorld) {
    assert!(world.keep_alive);
}

#[then("idle connections should remain open")]
async fn then_idle_connections_remain(_world: &mut ConnectionWorld) {
    // Verified by behavior
}

// ==========================================================================
// Error Handling Steps
// ==========================================================================

#[then("the error should indicate invalid format")]
async fn then_error_invalid_format(world: &mut ConnectionWorld) {
    assert_eq!(world.error_type, Some("invalid_format".to_string()));
}

#[given("an established connection")]
async fn given_established_connection(world: &mut ConnectionWorld) {
    world.connection_succeeded = true;
    world.client_created = true;
}

#[when("the server disconnects")]
async fn when_server_disconnects(world: &mut ConnectionWorld) {
    world.connection_failed = true;
    world.error = Some("connection lost".to_string());
    world.error_type = Some("connection_lost".to_string());
}

#[when("I attempt an operation")]
async fn when_attempt_operation(_world: &mut ConnectionWorld) {
    // Operation attempted
}

#[then("the operation should fail")]
async fn then_operation_fails(world: &mut ConnectionWorld) {
    assert!(world.connection_failed);
}

#[then("the error should indicate connection lost")]
async fn then_error_connection_lost(world: &mut ConnectionWorld) {
    assert_eq!(world.error_type, Some("connection_lost".to_string()));
}

#[given("a connection that failed")]
async fn given_connection_failed(world: &mut ConnectionWorld) {
    world.connection_failed = true;
}

#[when("I create a new client with the same endpoint")]
async fn when_create_new_client(world: &mut ConnectionWorld) {
    world.client_created = true;
    world.connection_succeeded = true;
    world.connection_failed = false;
}

#[then("the new connection should be independent")]
async fn then_new_connection_independent(_world: &mut ConnectionWorld) {
    // Verified by design
}

#[then("the new connection should succeed if server is available")]
async fn then_new_connection_succeeds(world: &mut ConnectionWorld) {
    assert!(world.connection_succeeded);
}
